### PR TITLE
#38 - Support zlib compression

### PR DESCRIPTION
**Pull request by rahul007ashok**
Added Gzip support for Fluent plugin kinesis

### CODE REVIEW

**Comment by riywo** [comment]
Hi @rahul007ashok ,

This is very interesting. I'm checking it out.

**Comment by rahul007ashok** [comment]
Hi @riywo ,

I was wondering if you have any questions or suggestions. Thanks

**Comment by riywo** [comment]
Hi @rahul007ashok ,

Could you confirm this contribution is released under Apache 2.0?

Best

**Comment by rahul007ashok** [comment]
Hi @riywo 
Yes, this contribution will be under Apache 2.0.
Rahul

**Comment by riywo** [comment]
Thank you for confirmation!

File: lib/fluent/plugin/out_kinesis.rb [hunk 1]
```diff
@@ -220,7 +222,11 @@ def get_key(name, record)
     end
 
     def build_data_to_put(data)
-      Hash[data.map{|k, v| [k.to_sym, v] }]
+      if @gzip_compression
+        Hash[data.map{|k, v| [k.to_sym, k=="data" ? Zlib::Deflate.deflate(v) : v] }]
```
Comment (riywo) [inline lib/fluent/plugin/out_kinesis.rb]: In `write` method, this plugin checks whether the size of record exceeds 1MB limitation or not. So, deflating after the checking will cause some trouble.

**Comment by riywo** [comment]
Hi @rahul007ashok ,

I've checked somehow and added a comment. Could you modify it to deflate before the checking size of records?

Also, could you add the description for `gzip_compression` option in README.md?

Best,
riywo

**Comment by rahul007ashok** [comment]
Hi @riywo ,

Added documentation to the README and also modified the write method to deflate before checking the size of records.

Thanks,
Rahul

**Comment by riywo** [comment]
@rahul007ashok 

Thanks for update. Let me check...

Could you rebase your changes to a single commit please?

**Comment by rahul007ashok** [comment]
@riywo 
Done.

**Comment by rahul007ashok** [comment]
@riywo 
I have modified the pull request to do zlib instead. Zlib compression is more widely used and is more efficient than gzip. Updated code, documentation and commit message to reflect this.

**Comment by riywo** [comment]
Thank you for the PR! I will release the new version later.

**Comment by wryun** [comment]
Don't really know anything about fluentd plugins, but why isn't this done in the format method instead? Then we save space buffering on disk as well (and waste less memory during processing).

**Comment by rahul007ashok** [comment]
The format method returns a MessagePack serialized array. MessagePack is not compatible with zlib/gzip(or other compression methods). So it's not possible for the format method to do the compression(although this would be better).

**Comment by wryun** [comment]
But the format method can also return a raw string, AFAIK.

Would make it somewhat harder to do, but not impossible? e.g. first 4 bytes of each record is length of record. Or you can get fancy and use some encoding like protobuf (i.e. anything which allows concatenation of binary outputs to make a valid sequence).

**Comment by wryun** [comment]
Hmm, also, why doesn't MessagePack 'support zlib/gzip'? Can't it just store the zipped field in the binary format? (ok, I should try this myself)

**Comment by riywo** [comment]
v0.4.0 is also released at rubygems.
https://rubygems.org/gems/fluent-plugin-kinesis/versions/0.4.0

Thanks!

**Pull request merged**